provider/metrics/container_insights.go

```diff
@@ -0,0 +1,315 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/pkg/errors"
+	"github.com/thoas/go-funk"
+	"github.com/virtual-kubelet/azure-aci/client/aci"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	stats "github.com/virtual-kubelet/virtual-kubelet/node/api/statsv1alpha1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// container insights implementation of podStatsGetter interface
+type containerInsightsPodStatsGetter struct {
+	metricsGetter        ContainerGroupMetricsGetter
+	resourceGroup        string
+	cumulativeUsageCache *cache.Cache
+}
+
+// some of values in stats.PodStats are require cumulative data. But Container Insights be only able to
+// provide the average value in 1 minute period. So we have to calculate the cumulative value from the average.
+// we need to remember the last cumulative value and add new value of next time windows.
+type podCumulativeUsage struct {
+	podUID                         string
+	containersUsageCoreNanoSeconds map[string]cumulativeUsage
+	networkRx                      cumulativeUsage
+	networkTx                      cumulativeUsage
+}
+
+type cumulativeUsage struct {
+	lastUpdateTime time.Time
+	value          uint64
+}
+
+func NewContainerInsightsMetricsProvider(metricsGetter ContainerGroupMetricsGetter, resourceGroup string) *containerInsightsPodStatsGetter {
+	return &containerInsightsPodStatsGetter{
+		metricsGetter:        metricsGetter,
+		resourceGroup:        resourceGroup,
+		cumulativeUsageCache: cache.New(time.Minute*30, time.Minute*30),
+	}
+}
+
+func (containerInsights *containerInsightsPodStatsGetter) getPodStats(ctx context.Context, pod *v1.Pod) (*stats.PodStats, error) {
```

Comment:
Please use single letter names for these objects.
e.g.

```
func (g *containerInsightsPodStatsGetter) getPodStats(ctx context.Context, pod *v1.Pod) (*stats.PodStats, error) {
```

We can probably also shorten `containerInsightsPodStatsGetter` to `insightsGetter` or maybe `getterV0`

Reply:
Because we don't need the name repeated everywhere.
If we put a doc string on the object like `// Implements PodStatsGetter for Container Insights` it puts that important information where the object is defined and reduces the noise everywhere else.
It's also just idiomatic go.

Reply:
Agree with you. I fixed the long name.